pkg/operator/reconciler.go

```diff
@@ -0,0 +1,402 @@
+				return false
+			}
+
+			return e.Object.GetName() == ClusterOperatorName
+		},
+	}
+
+	dsUpdatePredicate := predicate.Funcs{UpdateFunc: func(e event.UpdateEvent) bool {
+		if !validateUpdateEvent(&e) {
+			return false
+		}
+
+		dsOld := e.ObjectOld.(*appsv1.DaemonSet)
+		dsNew := e.ObjectNew.(*appsv1.DaemonSet)
+
+		return !equality.Semantic.DeepEqual(dsOld.Status, dsNew.Status) ||
+			!equality.Semantic.DeepEqual(dsOld.Spec, dsNew.Spec)
+	}}
+
+	tunedPredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			if e.Object == nil {
+				klog.Error("create event has no runtime object to create")
+				return false
+			}
+			tuned := e.Object.(*tunedv1.Tuned)
+			return tuned.Spec.ManagementState != ""
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if !validateUpdateEvent(&e) {
+				return false
+			}
+
+			tunedOld := e.ObjectOld.(*tunedv1.Tuned)
+			tunedNew := e.ObjectNew.(*tunedv1.Tuned)
+
+			return !reflect.DeepEqual(tunedOld.Spec.ManagementState, tunedNew.Spec.ManagementState)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			tuned := e.Object.(*tunedv1.Tuned)
+			return !e.DeleteStateUnknown && tuned.Spec.ManagementState != ""
+		},
+	}
+
+	profilePredicates := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			if e.Object == nil {
+				klog.Error("create event has no runtime object to create")
+				return false
+			}
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if !validateUpdateEvent(&e) {
+				return false
+			}
+			profileOld := e.ObjectOld.(*tunedv1.Profile)
+			profileNew := e.ObjectNew.(*tunedv1.Profile)
+			return !reflect.DeepEqual(profileOld.Status.Conditions, profileNew.Status.Conditions)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return !e.DeleteStateUnknown
+		},
+	}
+
+	err := ctrl.NewControllerManagedBy(mgr).
+		For(&configv1.ClusterOperator{}, builder.WithPredicates(cvoPredicates)).
+		Owns(&appsv1.DaemonSet{}, builder.WithPredicates(dsUpdatePredicate)).
+		Owns(&tunedv1.Tuned{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(
+			&source.Kind{Type: &tunedv1.Tuned{}},
+			handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+				return []reconcile.Request{
+					{NamespacedName: types.NamespacedName{
+						Name:      tunedv1.TunedClusterOperatorResourceName,
+						Namespace: ntoconfig.OperatorNamespace(),
+					}},
+				}
+			}),
+			builder.WithPredicates(tunedPredicates)).
+		Watches(
+			&source.Kind{Type: &tunedv1.Profile{}},
+			handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+				return []reconcile.Request{
+					{NamespacedName: types.NamespacedName{
+						Name:      tunedv1.TunedClusterOperatorResourceName,
+						Namespace: ntoconfig.OperatorNamespace(),
+					}},
+				}
+			}),
+			builder.WithPredicates(profilePredicates)).
+		Complete(r)
+	if err != nil {
+		return err
+	}
+	//TODO - not sure this is the right place
```

Comment:
Well, it definitely works as is, but I'm not sure it's logical. Why not just keep it in main.go?

Reply:
Because we need the server to first start and then register the version. 
This is not possible in the current implementation since after starting the manager :  
https://github.com/openshift/cluster-node-tuning-operator/blob/b922bddae5c88ba489020ba7f6b235fc13d00ec3/cmd/cluster-node-tuning-operator/main.go#L140
We cannot add additional functionality right after that command.

Reply:
@yanirq That is what I thought initially, but in this current location in reconciler.go `RegisterVersion(version.Version)` is being called before mgr.Start, so it would be equivalent to put it in main.go (before mgr.Start). 

I don't think RegisterVersion needs to be called after mgr.Start because it the metrics will be stored in the local prometheus registry until the server comes up, at which point they will be scraped by OCP monitoring. I did a few tests to confirm this.